### ISSUE TITLE
Freeze of picongpu if used MPI_Allgather in plugin

### ISSUE THREAD

**Issue by theZiz** [ComputationalRadiationPhysics/picongpu#1322]
Hi,

I noticed a bug in picongpu while writing at isaac. I managed to write a minimal example plugin, which shows the error. The plugin called "MPIStressTest" is avalaible at my picongpu fork in the branch "mpistresstest": https://github.com/theZiz/picongpu/tree/mpistresstest

The plugin uses this code in the `notify()` call:

``` C++
testClass.update( cellDescription );
//Some sample MPI magic
int numProc;
MPI_Comm_size(mpi_world, &numProc);
float buffer[numProc];
float my_value = float(rand());
if (mode & 1)
    MPI_Allgather( &my_value, 1, MPI_FLOAT, buffer, 1, MPI_FLOAT, mpi_world);
if (mode & 2)
    MPI_Bcast( buffer, numProc, MPI_FLOAT, 0, mpi_world);
```

The `mode` in line 7 and 9 is a parameter set in the submit cfg. The applications freezes if the mode is 1 or 3, so if "MPI_Allgather" is called. However it may be, that MPI_Bcast makes picongpu freeze, too, from time to time. I was just not able to repdroduce this.

However this MPI call is the reason for the freeze, but the application does not freeze at this point. It freezes in the `update` method of the dummy test class `testClass` with this implementation:

```
void update( MappingDesc *cellDescription )
{
    const SubGrid<simDim>& subGrid = Environment< simDim >::get().SubGrid();
    DataConnector &dc = Environment< simDim >::get().DataConnector(); 
    FieldE * pField = &(dc.getData< FieldE > (FieldE::getName(), true));
    DataSpace< simDim > guarding = SuperCellSize::toRT() * cellDescription->getGuardingSuperCells();
    //printf("get device data box\n");
    FieldE::DataBoxType dataBox = pField->getDeviceDataBox();
    //printf("done\n");
    FieldE::DataBoxType shifted = dataBox.shift( guarding );
}
```

The function `FieldE::DataBoxType dataBox = pField->getDeviceDataBox();` does not return (Checked with the surrounded printfs) for one node. However every else node will block at any later MPI call, too, because of the frozen rank.

I defined DEBUG_EVENTS in src/libPMacc/include/eventSystem/Manager.tpp to have some debug informations at this point. If the appliaction freezes this is printed to stdout:

```
[1,13]<stdout>:TaskReceive 1 2
[1,13]<stdout>:TaskReceiveMPI 1 2
```

Most of the time it happens right at the start, but sometimes after more than 100 steps, sometimes is runs the full simulation without problems. Most of the time the error is exactly as posted, but sometimes the 2 is switched for 3 or the message appears twice, but in this case the ranks in front of <stdout> are different.

I used the Laserwakefield example with default parameters. In my submit cfg I have only the test plugin activated. I am using 16 gpus with x=2, y=4, z=2 and a grid of 128 x 512 x 128. If I set x=1, y=16 and z=1 the bug appears more rarely, but still occurs.

It doesn't matter, how I start the job (direct qsub or qsub with bash).

If you need more informations about my test settings, don't hesitate to ask!

Greetings, Ziz

**Comment by ax3l** [comment]
Thx for the report and minimal example, that is excellent.

I think you are using `MPI_Allgather` wrong: the first argument (result) is also an array of the same size as the input in your example.

**Comment by theZiz** [comment]
I use it correctly. See this image http://mpitutorial.com/tutorials/mpi-scatter-gather-and-allgather/allgather.png from this page http://mpitutorial.com/tutorials/mpi-scatter-gather-and-allgather/ .

**Comment by ax3l** [comment]
yes, that was outdated but gh emails are going out so fast ;)

**Comment by theZiz** [comment]
I would say the gh mails have the right speed, you are clicking "Comment" to fast. :P

> Can you just use the communicator that we actually provide without MPI_Com_dup
Yes. It fails the same way. I also tried with MPI_WORLD earlier. Tbh I used a dup instead of MPI_WORLD in the very first way in isaac because of this strange bug. It didn't help, but I got told creating a dup is better™.

> Are you running on hypnos-hzdr?
Yes

> Are openib flags (openib.conf) correctly set for the mpiexec call (see .tpl file)?
They are set. I have no idea, whether they are correct. It's generated from the default example and contains:
´´´
btl_openib_rdma_pipeline_send_length = 100000000
btl_openib_rdma_pipeline_frag_size = 100000000
´´´

> does it affect the latest stable release, too or only dev?
I don't know. I will test this and report.

**Comment by ax3l** [comment]
> gh mails have the right speed,

yes ;)

> but I got told creating a dup is better™.

that is absolutely true; please just dup from `gc.getCommunicator().getMPIComm()` instead because that is the communicator of the simulation you interact with. (In case we split MPI_COMM_WORLD in the future for other usages.)

> I will test this and report.

thanks a lot

btt: I don't know right now what could be causing this.

**Comment by ax3l** [comment]
latest idea (offline): use `false` in `FieldE * pField = &(dc.getData< FieldE > (FieldE::getName(), true));` to force sync with device.

**Comment by psychocoderHPC** [comment]
The problem is that all blocking MPI commands are not notify the event system:

B = begin
E = end
```
[1,1]<stdout>:B notify
[1,2]<stdout>:B notify
[1,2]<stdout>:EField updated
[1,2]<stdout>:TaskSend state=1  ( state 1 is copy date from device to host, the mpi send is not started at the moment)
[1,2]<stdout>:TaskCopyDeviceToHost
[1,2]<stdout>:B MPI_Allgather  # rank 2 enter blocking mpi function before all PMacc mpi function are finished (can not finished if event loop is not called)
[1,1]<stdout>:TaskReceive state=1 exchange=7
[1,1]<stdout>:TaskReceiveMPI  (this taks wait for the mpi send from rank 2)
```

**Comment by theZiz** [comment]
I tried @ax3l 's suggestions with `getData<...>(..., false);` and it worked for my incredible big amount of 1 test. So I am still unsure, whether this fixes the issue, but we still can hope. ;)

@psychocoderHPC So ax3l's suggestion would fix this? Or what do I have to do?

**Comment by ax3l** [comment]
@theZiz the problem is a bit deeper and @psychocoderHPC found it today. We discussed it but since it's late Friday the fix will not come in today.

the underlying problem is that our event system loop is not threaded but synchronously, *in line/interleaved* executed with the main program (that's because, e.g., MPI + threading is a topic of it's own.)

We have a specific event construction in the code that is totally correct in data dependencies but causes a deadlock as soon as a blocking operation starts while an other half-open operation (that did not need to finish before the collective) is open. As @psychocoderHPC just told me offline, we have some cuda operations that depend on an mpi operation and we do forward the sub-state of that mpi operation since the data dependency is fulfilled but the previous mpi op can still create other mpi events (that do not influence the data of the cuda op behind it).

its actually possible that forcing a sync might work-around this until monday because the sync might wait for all mpi events to be finished on the device. but could also just move the problem away, it's a race condition.

**Comment by ax3l** [comment]
and yeah @theZiz you reproducibly triggered a logic error construction in our core event handling, that's not often the case and really awesome! :sparkles:

**Comment by ax3l** [comment]
P.S.: a reliable work-around would be until monday: add a `Environment<>::get().Manager().waitForAllTasks();` before doing your MPI call.

**Comment by bussmann** [comment]
Awesome work by everybody!

**Comment by theZiz** [comment]
Thanks, but I think I can wait until monday. ;)

**Comment by ax3l** [comment]
okok, just one question:
![yunopatch](https://cloud.githubusercontent.com/assets/1353258/12206721/c53eb23c-b642-11e5-84bd-6ca20d4acda3.jpg)

just kidding, have a nice weekend :)

**Comment by theZiz** [comment]
Not's like I don't have anything else to do about my thesis.... Like... Writing it. :P

Have a nice weekend, too. ;)

**Comment by ax3l** [comment]
hm, or maybe
![notweekend](https://cloud.githubusercontent.com/assets/1353258/12207508/45b8334e-b647-11e5-97dd-23113430d3e4.jpg)

**Comment by psychocoderHPC** [comment]
You can remove your sync workaround.
The real fix is to change [this line](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/libPMacc/include/memory/buffers/GridBuffer.hpp#L508-L510)

to

```C++
evR += asyncSend(serialEvent, sendEx, copyEvent);
```

**Comment by ax3l** [comment]
closed in `0.1.2` and `dev`. @psychocoderHPC do you want to close?

**Comment by ax3l** [comment]
closed via #1326
